### ISSUE TITLE
Dart imports are not case sensitive

### ISSUE THREAD

**Issue by AbhishekDoshi26** [dart-lang/linter#3244]

**Comment by srawlins** [comment]
Ah I see. Yes then a lint rule makes sense. Something you can opt-in to, understanding what it means.

**Comment by lrhn** [comment]
It's probably outside the analyzer's scope to lint that file names are actually lower-case (because how do you even check).

The Pub tool could perhaps check that you only upload lower-case files. IIRC, it gets a zip file of the package directory, so it can check the file names in the zip without going through a file system. (@jonasfj)

**Comment by pq** [comment]
> The Pub tool could perhaps check that you only upload lower-case files. 

Interestingly, [file_names](https://dart-lang.github.io/linter/lints/file_names.html) should flag these and is part of the core lint set.  I'd expect these pub packages are already getting their scores dinged...

**Comment by lrhn** [comment]
Will the `file_names` lint also check directory names, or is it only library names? (The documentation only mentions library names, and probably even only library names matching `\w+.dart`)
The problem described in this issue is a directory name.

I'd want a warning if any part of an import path contains an upper-case character (I don't care about `-` vs `_` at all, those work on all platforms), and I'd like a Pub warning if I try to upload any file or directory with an upper-case character in it.

**Comment by pq** [comment]
> Will the file_names lint also check directory names, or is it only library names?

Ah! Sorry, I wasn't looking closely.  `file_names` does not check directories so it's not sufficient.  We could definitely consider broadening it to look at the package-relative path.  It'd be interesting to scan some pub packages (and Google3) and see how disruptive this change might be.

>  I'd like a Pub warning if I try to upload any file or directory with an upper-case character in it.

I'd be curious to get @jonasfj's thoughts but sounds good to me.

**Comment by AbhishekDoshi26** [comment]
@pq just a question, why just. the packages? Can the lint rules throw warnings even for the local folders? Because if I have a file as `screens/home.dart` and if I import it as `Screens/home.dart` it works on macOS. But when such imports are moved into pipelines, it breaks and gets really difficult to debug.

**Comment by pq** [comment]
> just a question, why just. the packages? 

Sorry, I wasn't too clear.  I just meant that we'd not want to check **absolute** path (which was probably obvious).  Definitely, I think we'd want to flag `Screens/home.dart` locally.  If we pile this onto `file_names` this would get reported on `home.dart` btw and not on the local import of it...

**Comment by jonasfj** [comment]
> I'd like a Pub warning if I try to upload any file or directory with an upper-case character in it.

We did a while ago forbid uploads when package contains two files with the same name (case-insensitive), because extracting such packages on MacOS and Windows is problematic.

I think it would be reasonably to warn against upper-case letters in Dart file names when publishing (ie. only for `*.dart` files).
But it can only be a warning, since people might be need to include files for testing reasons.

IMO, we could make **hint** in the analyzer that discourages imports when the casing doesn't match the casing on the file-system. Afterall Windows and MacOS are typically case-preserving, and if your import-statement doesn't match the casing it'll cause problems on Linux, and generally indicates that you've made a bug.

It would also be preferable to surface this early, rather than only surfacing it packages are publishing. This could still occur within a single repository.

**Comment by lrhn** [comment]
The problem is in detecting the case of the file in a case-insensitive file system.

Asking if a file exists at another capitalization is not sufficient, the other capitalization could exist on a case-sensitive file system.

You'd probably have to:
* Check if an imported file exists.
* If so, list the files of the directory, presumably getting the stored case of the files, and check if there is no file with the same capitalization as the import.
* Since directories have cases too, repeat that for each directory all the way up to the root, to check that the entire path has the same casing as the import URI.

That's a lot of extra work.

Maybe there is a way to have the system canonicalize an arbitrary path (we'd [want that](https://github.com/dart-lang/path/issues/102) anyway). I don't think it's trivial (https://stackoverflow.com/questions/7195337/how-do-i-get-a-path-with-the-correct-canonical-case-in-powershell), and we'd need to do it on each supported platform, or at least those where we don't know [for sure](https://man7.org/linux/man-pages/man5/ext4.5.html#:~:text=feature%20be%20enabled.-,casefold,-This%20ext4%20feature) that all file systems are case sensitive. And in a way where canonicalization doesn't also resolve symbolic links. Also, Unicode!

**Comment by kissa-hh** [comment]
Just adding a case that might highlight the potential hazards of this issue.

We had `import "services/services.dart;` in our code base, and someone accidentally added `import "services/Services.dart";`. Suddenly we had two `Services` classes in our runtime, and all of our services (which were static members in thet class) were instantiated twice. It took our team quite a while to realize what was going on and even then we found it hard to believe that a static member could be initialized twice. I can imagine that a bug like that could have serious consequences. 

I'd prefer this to be solved at the language level, not as a linter rule. My preferred solutions:
- prevent importing "File.dart" if "File.dart" is not the canonical filename (but, as mentioned, this might be hard to implement; [the C standard library also doesn't seem to have an easy way to find out the canonical filename](https://stackoverflow.com/questions/73463994/how-to-get-canonical-filename-on-a-case-insensitive-filesystem))
- prevent importing "File.dart" and "file.dart" in the same program
- make importing "File.dart" and "file.dart" the same thing (by lowercasing all filenames before registering them)

**Comment by kissa-hh** [comment]
Here's a test case to show what was going on:

main.dart:
```
import 'Test.dart' as t1;
import 'test.dart' as t2;

void main() {
	var n = t1.Test.x;
	var n2 = t2.Test.x;
}
```

test.dart:
```
import 'Test.dart' as t1;
import 'test.dart' as t2;

void main() {
	var n = t1.Test.x;
	var n2 = t2.Test.x;
}
```

Result:
```
f() called!
f() called!
```

(Expected result: that program would either produce an compile error or just print the message once.)

**Comment by sorewanya** [comment]
I have similar problem in imports:
file structure: 
```
this_page.dart ; this_entity.dart ; this_entity.g.dart; 
that_entity.dart ; that_entity.g.dart ; 
```
this_page.dart:
```
import 'this_entity.dart';
class ThisPage ...
...
```

1)copy **this_page.dart >> that_page.dart** 
2)in **"that_page.dart"** using "select all" (ctrl+F2 in VSC) i replace **"This" >> "That"** in classes name, but forget, what it real "select ALL" - imports lines is ALL too. 

result of this: 
Dart don't find copyWith() extension from that_entity.g.dart file. 
`The method 'copyWith' isn't defined for the type 'ThatEntity'.`

the ThatEntity itself show "package:packname/**T**hat_entity.dart", 

solution:
T=>t in import

I think its not feature, but cant named bug ))
